cluster-api/cloud/google/machinesetup/config_types.go

```diff
@@ -0,0 +1,126 @@
+package machinesetup
+
+import (
+	"fmt"
+	"github.com/ghodss/yaml"
+	"io"
+	"io/ioutil"
+	clustercommon "k8s.io/kube-deploy/cluster-api/pkg/apis/cluster/common"
+	clusterv1 "k8s.io/kube-deploy/cluster-api/pkg/apis/cluster/v1alpha1"
+	"k8s.io/kube-deploy/cluster-api/util"
+	"os"
+)
+
+type ConfigWatch struct {
+	Path string
+}
+
+type ValidConfigs struct {
+	configList *configList
+}
+
+type configList struct {
+	Items []config `json:"items"`
+}
+
+type config struct {
+	// A list of the valid combinations of ConfigParams that will
+	// map to the given Image and Metadata.
+	Params []ConfigParams `json:"machineParams"`
+
+	// This can either be a full projects path to an image/family,
+	// or just the image name which is in the project.
+	// If it's an image in the project, this field may be
+	// identical to the OS field (see GCEClient.getImagePath()).
+	Image    string   `json:"image"`
+	Metadata Metadata `json:"metadata"`
+}
+
+type Metadata struct {
+	StartupScript string `json:"startupScript"`
+}
+
+type ConfigParams struct {
+	OS       string
+	Roles    []clustercommon.MachineRole
+	Versions clusterv1.MachineVersionInfo
+}
+
+func NewConfigWatch(path string) (*ConfigWatch, error) {
+	if _, err := os.Stat(path); err != nil {
+		return nil, err
+	}
+	return &ConfigWatch{Path: path}, nil
+}
+
+func (cw *ConfigWatch) ValidConfigs() (*ValidConfigs, error) {
+	file, err := os.Open(cw.Path)
+	if err != nil {
+		return nil, err
+	}
+	return parseMachineSetupYaml(file)
+}
+
+func parseMachineSetupYaml(reader io.Reader) (*ValidConfigs, error) {
+	bytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	configList := &configList{}
+	err = yaml.Unmarshal(bytes, configList)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ValidConfigs{configList}, nil
+}
+
+func (vc *ValidConfigs) GetYaml() (string, error) {
+	bytes, err := yaml.Marshal(vc.configList)
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+func (vc *ValidConfigs) GetImage(params *ConfigParams) (string, error) {
+	machineSetupConfig, err := vc.matchMachineSetupConfig(params)
+	if err != nil {
+		return "", err
+	}
+	return machineSetupConfig.Image, nil
+}
+
+func (vc *ValidConfigs) GetMetadata(params *ConfigParams) (Metadata, error) {
+	machineSetupConfig, err := vc.matchMachineSetupConfig(params)
+	if err != nil {
+		return Metadata{}, err
+	}
+	return machineSetupConfig.Metadata, nil
+}
+
+func (vc *ValidConfigs) matchMachineSetupConfig(params *ConfigParams) (*config, error) {
+	for _, conf := range vc.configList.Items {
+		for _, validParams := range conf.Params {
+			if params.OS != validParams.OS {
+				continue
+			}
+			foundRoles := true
+			for _, role := range params.Roles {
+				if !util.RoleContains(role, validParams.Roles) {
```

Comment:
This ensures the matched config has at least all the roles in the param. However, it still matches even if the config has additional roles the param did not have.

Reply:
Perhaps cover this in your unit tests?

Reply:
I added a case to the test for it.

Reply:
I originally planned to allow the configs to have more roles than the params because I intended for the config to support any combination of roles that it lists there. It would consolidate the configs so if you had multiple roles that can all use the same script, you wouldn't need a config for each combination. But I thought about it some more and it's kind of hard to reason about this situation since we only have a master and node role. I switched it to ensure it's an exact match to avoid confusion.

Reply:
I modified the check to account for this and added another case to the test.